File: structure/heap.php
<?php

class Heap
{
    private $a = [];
    private $n;
    private $count;

    public function __construct($capacity = 10)
    {
        $this->n = $capacity;
        $this->count = 0;
    }

    public function insert($data)
    {
        if ($this->count >= $this->n) {
            return false;
        }
        $this->count++;
        $this->a[$this->count] = $data;
        $i = $this->count;
        while (floor($i / 2) > 0 && $this->a[floor($i / 2)] > $this->a[$i]) {
            $temp = $this->a[$i];
            $this->a[$i] = $this->a[floor($i / 2)];
            $this->a[floor($i / 2)] = $temp;
            $i = floor($i / 2);
        }
        return true;
    }

    public function remove()
    {
        if ($this->count == 0)
            return false;
        $removeData = $this->a[1];
        $this->a[1] = $this->a[$this->count];
        $this->count--;
        $i = 1;  // 堆顶元素
        while (true) {
            $maxPos = $i;
            if ($i * 2 <= $this->count && $this->a[$i * 2] > $this->a[$i]) {
                $maxPos = 2 * $i;
            }
            if ($i * 2 + 1 <= $this->count && $this->a[$i * 2 + 1] > $this->a[$maxPos]) {
                $maxPos = 2 * $i + 1;
            }
            if ($maxPos == $i) {
                break;
            }
            $temp = $this->a[$i];
            $this->a[$i] = $this->a[$maxPos];
            $this->a[$maxPos] = $temp;
            $i = $maxPos;
        }
        return $removeData;
    }

    public function __toString()
    {
        return json_encode(array_values($this->a));
    }
}

$heap = new Heap();
$heap->insert(1);
$heap->insert(2);
$heap->insert(3);
$heap->insert(4);
$heap->insert(5);
$heap->insert(6);
$heap->insert(7);
echo $heap;

File: structure/Heap.java
import java.util.Stack;

/**
 * Heap
 */
public class Heap
{
    private int[] data;
    private int size;
    private int used;
    private boolean resize;

    public Heap(int capacity) {
        this(capacity, false);
    }

    public Heap(int capacity, boolean resize) {
        this.data = new int[capacity];
        this.size = capacity;
        this.used = 0;
        this.resize = resize;
    }

    public boolean push(int value) {
        if (used == size) return false;
        data[++used] = value;
        int index = used;
        while ((index >>> 1) > 0 && compare(value, data[index >>> 1]) > 0) {
            swap(index, index >>> 1);
            index = index >>> 1;
        }
        if (used == size - 1 && resize) expand();
        return true;
    }

    public int pop() {
        if (used == 0)
            throw new ArrayIndexOutOfBoundsException(0 + " >= " + used);
        int result = data[1];
        data[1] = data[used--];
        for (int point = 1;;) {
            int next = point, left = point << 1, right = (point << 1) + 1;
            if (left <= used && compare(data[left], data[next]) > 0)
                next = left;
            if (right <= used && compare(data[right], data[next]) > 0)
                next = right;
            if (next == point) break;
            swap(point, next);
            point = next;
        }
        return result;
    }

    @Override
    public String toString() {
        StringBuilder sbl = new StringBuilder().append("[Element: ").append(used).append(", Data: ");
        sbl.append(data[1]);
        for (int i = 2; i <= used; i++) {
            sbl.append(", ").append(data[i]);
        }
        sbl.append("]");
        return sbl.toString();
    }

    protected int compare(int num1, int num2) {
        return num1 - num2;
    }

    private void expand() {
        int oldSize = size;
        int newSize;
        if (Integer.MAX_VALUE - oldSize <= (oldSize >>> 1))
            newSize = Integer.MAX_VALUE;
        else
            newSize = oldSize + oldSize >>> 1;
        int[] newData = new int[newSize];
        System.arraycopy(data, oldSize, newData, 0, oldSize);
        size = newSize;
        data = newData;
    }


    private void swap(int index1, int index2) {
        int temp = data[index1];
        data[index1] = data[index2];
        data[index2] = temp;
    }
}
